src/operator/nn/dropout-inl.h

```diff
@@ -95,46 +73,75 @@ struct DropoutParam : public dmlc::Parameter<DropoutParam> {
 template<typename xpu, typename DType>
 class DropoutOp : public Operator {
  public:
+  /*!
+   * \brief Dropout kernel, compute dropout tensor
+   */
+  struct DropoutKernel {
+    /*!
+     * \brief Dropout kernel function
+     * \param id Thread number (0-based representing count)
+     * \param gen Random number generator
+     * \param N Total number of items in the output
+     * \param step Step between items, related to parallelism
+     * \param dropout_out Output dropout values
+     * \param mask_out  Output mask (is multiplied to create dropout output, may be 0)
+     * \param input_data Input data to perform the dropout on
+     * \param pkeep Dropout rate (keep when the generated random number is less than this value)
+     */
+    MSHADOW_XINLINE static void Map(int id,
+                                    RandGenerator<xpu, DType> gen,
```

Comment:
should gen be by ref?

Reply:
No.

Reply:
@larroy it has to be passed by value since it will be a call into gpu kernel.

Reply:
@larroy, there's also another reason not to pass by reference here, can you guess what it is? :)

Reply:
@yzhliu thanks for the answer. I see that RNG_KERNEL_LOOP uses it by reference. I don't fully get the reason yet. The __device__ function you mean is RandGenerator<xpu,GType>::Impl? Just trying to understand.